appinventor/blocklyeditor/src/workspace_svg.js

```diff
@@ -632,9 +665,60 @@ Blockly.WorkspaceSvg.prototype.customContextMenu = function(menuOptions) {
     var comparatorA = comparisonName(a).toLowerCase();
     var comparatorB = comparisonName(b).toLowerCase();
 
-    if (comparatorA < comparatorB) return -1;
-    else if (comparatorA > comparatorB) return +1;
-    else return 0;
+    if (a.category != b.category){
+      return comparatorA.localeCompare(comparatorB);
+    }
+
+    // Sort by Category First, also handles other floating blocks
+    if (a.category == b.category && a.category != "Component"){
+      // Remove '1,'
+      comparatorA = comparatorA.substr(2);
+      comparatorB = comparatorB.substr(2);
+      return compareStrTextNum(comparatorA, comparatorB);
+    }
+
+    // 3.Component event handlers, lexicographically sorted by 
+    // type name, instance name, then event name
+    if (a.category == "Component" && b.category == "Component" && a.eventName && b.eventName){
+      if (a.typeName == b.typeName){
+        return compareStrTextNum(a.instanceName, b.instanceName);
+      }
+      return comparatorA.localeCompare(comparatorB);
+    }
+
+    // 4. For Component blocks, sorted internally first by type, 
+    // whether they are generic (generics precede specifics), 
+    // then by instance name (for specific blocks), 
+    // then by method/property name.
+    if (a.category == "Component" && b.category == "Component"){
+      var geneA = ',2';
+      if (a.isGeneric){
+        geneA = ',1';
+      }
+
+      var geneB = ',2';
+      if (b.isGeneric){
+        geneB = ',1';
+      }
+
+      var componentA = a.type + geneA;
+      var componentB = b.type + geneB;
+
+      var res = componentA.localeCompare(componentB);
+      if (res == 0){
+        // compare instance names
+        res = compareStrTextNum(a.instanceName, b.instanceName);
```

Comment:
With "any component" blocks, this can result in an uncaught exception because instanceName will be undefined and then in compareStrTextNum you dereference undefined to execute the regular expression.